tools/syz-declextract/run.go

```diff
@@ -62,35 +64,91 @@ func main() {
+			case *ast.Struct:
+				netlinks = append(netlinks, node.(*ast.Struct))
+			case *ast.Include:
+				includes = append(includes, node.(*ast.Include))
 			}
-			allOut = append(allOut, renameSyscall(line, syscallNames)...)
 		}
 	}
+
 	close(files)
-	writeOutput(allOut, *outFile)
+	writeOutput(includes, syscalls, netlinks, *outFile)
 }
 
-func writeOutput(allOut []string, outFile string) {
-	slices.Sort(allOut)
-	allOut = slices.CompactFunc(allOut, func(a string, b string) bool {
-		// We only compare the part before "$" for cases where the same system call has different parameter names,
+func writeOutput(includes []*ast.Include, syscalls []*ast.Call, netlinks []*ast.Struct, outFile string) {
+	slices.SortFunc(includes, func(a, b *ast.Include) int {
+		return strings.Compare(a.File.Value, b.File.Value)
+	})
+	includes = slices.CompactFunc(includes, func(a, b *ast.Include) bool {
+		return a.File.Value == b.File.Value
+	})
+
+	slices.SortFunc(syscalls, func(a, b *ast.Call) int {
+		return strings.Compare(a.Name.Name, b.Name.Name)
+	})
+	syscalls = slices.CompactFunc(syscalls, func(a, b *ast.Call) bool {
+		// We only compare the the system call names for cases where the same system call has different parameter names,
 		// but share the same syzkaller type. NOTE:Change when we have better type extraction.
-		return strings.Split(a, "$")[0] == strings.Split(b, "$")[0]
+		return a.Name.Name == b.Name.Name
+	})
+	slices.SortFunc(netlinks, func(a, b *ast.Struct) int {
+		return strings.Compare(a.Name.Name, b.Name.Name)
 	})
+
+	autoGeneratedNotice := "# Code generated by syz-declextract. DO NOT EDIT."
+	mmap2 := "_ = __NR_mmap2\n"
+	eh := ast.LoggingHandler
+	desc := ast.Parse([]byte(autoGeneratedNotice), "", eh)
+	for _, node := range includes {
+		desc.Nodes = append(desc.Nodes, node)
+	}
+	for _, node := range syscalls {
+		desc.Nodes = append(desc.Nodes, node)
+	}
+	desc.Nodes = append(desc.Nodes, ast.Parse([]byte(mmap2), "", eh).Nodes...)
+	var netlinkNames []string
+	for _, node := range netlinks {
+		desc.Nodes = append(desc.Nodes, node)
+		netlinkNames = append(netlinkNames,
+			"\tpolicy"+strconv.Itoa(len(netlinkNames))+" msghdr_auto["+node.Name.Name+"]\n")
+	}
+	netlinkUnion := `type msghdr_auto[POLICY] msghdr_netlink[netlink_msg_t[autogenerated_netlink, genlmsghdr, POLICY]]
+resource autogenerated_netlink[int16]
+syz_genetlink_get_family_id$auto(name ptr[in, string], fd sock_nl_generic) autogenerated_netlink
+sendmsg$autorun(fd sock_nl_generic, msg ptr[in, auto_union], f flags[send_flags])
+auto_union [
+` + strings.Join(netlinkNames, "") + "]"
+	netlinkUnionParsed := ast.Parse([]byte(netlinkUnion), "", eh)
+	if netlinkUnionParsed == nil {
+		tool.Failf("Parsing Error")
+	}
+	desc.Nodes = append(desc.Nodes, netlinkUnionParsed.Nodes...)
+
 	err := os.WriteFile(outFile,
-		[]byte("# Code generated by syz-declextract. DO NOT EDIT.\n"+strings.Join(allOut, "\n")+"\n_ = __NR_mmap2\n"), 0666)
+		ast.Format(desc),
```

Comment:
I was wrong that it will fix formatting.
We add/remove new lines during parsing:
https://github.com/google/syzkaller/blob/master/pkg/ast/parser.go#L21-L40
rather than during serialization.

The simplest fix for now will be to serialize, and then parse and serialize again (with a comment as to why we do such strange thing).

Reply:
Done!